service/src/main/java/uk/gov/mca/beacons/api/registration/application/RegistrationService.java

```diff
@@ -233,15 +233,32 @@ private void deleteAssociatedAggregates(
     }
   }
 
+  // we can make this more efficient because currently it does findByHexIdAndAccountHolderEmail loads of times
+  // can we just claim from here rather than looking it up again
   private void claimLegacyBeacon(Beacon beacon) {
     AccountHolder accountHolder = accountHolderService
       .getAccountHolder(beacon.getAccountHolderId())
       .orElseThrow(ResourceNotFoundException::new);
 
-    legacyBeaconService.claimByHexIdAndAccountHolderEmail(
+    List<LegacyBeacon> legacyBeacons = legacyBeaconService.findByHexIdAndAccountHolderEmail(
       beacon.getHexId(),
       accountHolder.getEmail()
     );
+
+    for (LegacyBeacon legacyBeacon : legacyBeacons) {
+      if (legacyBeacon.getRecoveryEmail() != null) {
```

Comment:
A few things on this. 

Shouldn't need to loop here as it loops within the claimBy methods. 


Unless you move the loop up and try to claimBeaconByRecoveryEmail, then if that doesn't match anything, then try and claim by legacy beacon owner email. 

Looking at this again, looks like the logic is also a bit off:


We have: accountHolderEmail (The email address we now want the beacon to be under)
legacyBeacon.email (The historical email address on the beacon)
legacyBeacon.recoveryEmail (The editable match email address)


We want to:
{

if(canClaim(legacyBeacon, accountHolder.email)){
    doTheClaim(legacyBeacon, accountHolder.email);
    return claimedBeacon;
}

return null;

}

canClaim(legacyBeacon, accountHolderEmail){
    var canClaim = false;
    
    if(legacyBeacon.recoveryEmail != null){
        canClaim = doesMatch(legacyBeacon.recoveryEmail, accountHolderEmail)
    }
    
    return canClaim || doesMatch(legacyBeacon.ownerEmail, accountHolderEmail)
}

doesMatch(email1, email2){
 return email1 ==email2; //poor matching, but example.
}

Reply:
@Evie-Skinner18 - Have you had a think about this?

Reply:
Just checking have you seen commit 5d767e89e06700c972abc6cba10e2b10395e1701 from 9th May? I actually got rid of the claimBy() methods

Reply:
Soz do you mean that the logic is still off since 9th May?

Reply:
Oh right we need to swap it round so it checks recoveryEmail first, then ownerEmail if no match on recoveryEmail..?